consul/client.go

```diff
@@ -337,6 +349,60 @@ func (c *Client) RPC(method string, args interface{}, reply interface{}) error {
 	return nil
 }
 
+// RPCDeadline is used to forward an RPC call to a consul server and abort on
+// timeout.
+func (c *Client) RPCDeadline(method string, args interface{}, reply interface{}, deadline time.Time) error {
+	server := c.servers.FindServer()
+	if server == nil {
+		return structs.ErrNoServers
+	}
+
+	// Forward to remote Consul
+	if err := c.connPool.RPCDeadline(c.config.Datacenter, server.Addr, server.Version, method, args, reply, deadline); err != nil {
+		c.servers.NotifyFailedServer(server)
+		c.logger.Printf("[ERR] consul: RPC failed to server %s: %v", server.Addr, err)
+		return err
+	}
+
+	return nil
+}
+
+// RPCRetry retries a given RPC until the function has exceeded the deadline
+// timer or returned a successful reply.  The number of retries is determined
+// based on the number of Consul servers at the start of the RPC.
+func (c *Client) RPCRetry(method string, args interface{}, reply interface{}, slaDeadline time.Time) (err error) {
+	// If an RPC fails to return in the per-server RPC timeout, the
+	// server is marked as failed (rotated it to the end of the server
+	// list and we try with the next available server hoping it meets our
+	// SLA).
+	now := time.Now()
+	numRetries := lib.MaxInt(c.servers.NumServers(), rpcQueryMaxServers) * rpcNumRetriesPerServer
+	perServerTimeout := time.Duration(slaDeadline.Sub(now).Nanoseconds() / int64(numRetries))
+
+	// Bound the per-server timeout.
+	//
+	// FIXME(sean@): Do these timeouts need to be adjusted based on
+	// whether or not the RPC is destined to the local datacenter or a
+	// remote datacenter?  Based on command/agent/rpc_client.go, it
+	// doesn't look like it.
+	if perServerTimeout < rpcMinRetryDuration {
+		perServerTimeout = rpcMinRetryDuration
+	} else if perServerTimeout > clientRPCConnMaxIdle {
+		// Clamp the max timeout to our max idle
+		perServerTimeout = clientRPCConnMaxIdle
+	}
+
+	// Loop until success or the SLA has been exceeded
+	for i := numRetries; now.Before(slaDeadline) && i > 0; i-- {
```

Comment:
I think this algorithm will overshoot the SLA by `perServerTimeout`. Probably want to calculate the next deadline and make sure that's less than `slaDeadline`.